### PR TITLE
Include templating support

### PR DESCRIPTION
**Pull request by StopMotionCuber**
This PR includes support for templating.
This means it can build different versions of the image with the same file, which can be configured via `buildinfo.json`.

Note that it is still missing logic to push tags, as well as the adaption of the watchdog.

### CODE REVIEW

**Comment by Fank** [comment]
I will take care about the watchdog, if this got merged

**Comment by StopMotionCuber** [comment]
> I will take care about the watchdog, if this got merged

It would be really great if you could take care of that, I'm not really proficient in Go and my PR would likely get ugly.
However, from my point of view both repos are tightly coupled and merging this PR would break the watchdog, which is why I would like to ask you to start working on the watchdog before this gets merged.

Furthermore these were the requirements I had in mind for the watchdog:
 - Write the new `buildinfo.json` (obviously)
 - Include proper support for the `stable` tag and don't alias it to `latest` (which is why I made this PR in the first place)
 - Adapt README.md modifications. Here I included comments `<!-- start autogeneration tags -->` and `<!-- end autogeneration tags -->` which should be markers for those parts which get generated by the watchdog.

It would be great if you take these into account. Please tag me when you are finished, I'll happily review your PR.

File: README.md [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+<!-- start autogeneration tags -->
 * `1.1.26`, `1.1`, `latest`, `stable` [(1.1/Dockerfile)](https://github.com/factoriotools/factorio-docker/blob/master/1.1/Dockerfile)
 * `1.0.0`, `1.0` [(1.0/Dockerfile)](https://github.com/factoriotools/factorio-docker/blob/master/1.0/Dockerfile)
 * `0.18.47`, `0.18` [(0.18/Dockerfile)](https://github.com/factoriotools/factorio-docker/blob/master/0.18/Dockerfile)
 * `0.17.79`, `0.17` [(0.17/Dockerfile)](https://github.com/factoriotools/factorio-docker/blob/master/0.17/Dockerfile)
 * `0.16.51`, `0.16` [(0.16/Dockerfile)](https://github.com/factoriotools/factorio-docker/blob/master/0.16/Dockerfile)
 * `0.15.40`, `0.15` [(0.15/Dockerfile)](https://github.com/factoriotools/factorio-docker/blob/master/0.15/Dockerfile)
 * `0.14.23`, `0.14` [(0.14/Dockerfile)](https://github.com/factoriotools/factorio-docker/blob/master/0.14/Dockerfile)
+<!-- end autogeneration tags -->
+
```
Comment (SuperSandro2000) [inline README.md:12]:
```suggestion
<!-- end autogeneration tags -->
```

File: buildinfo.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "1.1.26": {
+    "sha1": "5736da3d068ee74b282926b99ad0255368569b51",
+    "tags": ["1.1.26", "1.1", "latest"]
+  },
+  "1.0.0": {
+    "sha1": "a50dd6f1ab17acde5a8d77bc5fb478c798d9c20e",
+    "tags": ["1.0.0", "1.0"]
+  }
+}
```
Comment (SuperSandro2000) [inline buildinfo.json:10]: Missing final new line.

File: build.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    dockerfile_content = template.render(sha1=sha1, version=version)
+    with open(os.path.join(build_dir, "Dockerfile"), "w") as dockerfile:
+        dockerfile.write(dockerfile_content)
+    build_command = ["docker", "build", "."]
+    for tag in tags:
+        build_command.extend(["-t", f"factoriotools/factorio:{tag}"])
+    try:
+        subprocess.run(build_command, cwd=build_dir, check=True)
+    except subprocess.CalledProcessError:
+        print("Build of image failed")
```
Comment (SuperSandro2000) [inline build.py:30]: Does this swallow the error log?

File: buildinfo.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "1.1.26": {
+    "sha1": "5736da3d068ee74b282926b99ad0255368569b51",
+    "tags": ["1.1.26", "1.1", "latest"]
+  },
```
Comment (SuperSandro2000) [inline buildinfo.json:5]: We are missing the stable tag here.

File: build.py (hunk 3 above)
Comment (StopMotionCuber) [reply build.py:30]: Nope, see https://github.com/StopMotionCuber/factorio-docker/runs/2043351161

File: buildinfo.json (hunk 4 above)
Comment (StopMotionCuber) [reply buildinfo.json:5]: Yeah, I didn't want to put the whole buildinfo in manually as the (yet-to-be-updated) watchdog should generate this file

File: README.md (hunk 1 above)
Comment (StopMotionCuber) [reply README.md:12]: This... is not a change?

Comment (SuperSandro2000) [reply README.md:12]: GitHub is just not smart enough to render the removal of the extra new line.

File: README.md [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
 
 ## Tag descriptions
```
Comment (SuperSandro2000) [inline README.md:14]:
```suggestion
<!-- end autogeneration tags -->

## Tag descriptions
```

**Comment by StopMotionCuber** [comment]
Should be ready now. Tell me if you want to have some squashed commits as my history is a bit messy.

**Comment by SuperSandro2000** [comment]
I just squash it when the updater part is cleared.

**Review by Fank — APPROVED** [review]

**Comment by perobertson** [comment]
Was there a reason why jinja2 templating was used to generate a Dockerfile with the sha1 and version added in? From what I can tell it would keep the layer caching for all the previously published images. If that is not a concern, then docker build-args could be used instead to populate the env variables:
```Dockerfile
ARG VERSION
ENV VERSION=${VERSION}
```

This would result in the jinja code not being needed.

**Comment by Fank** [comment]
I see, i think i missed that, could you try to generate the image by using build-args?

**Comment by StopMotionCuber** [comment]
The reason is mostly that templating was the easiest solution that came to my mind for solving that mess we had before. Docker build-args also work and seem like the cleaner solution here, so feel free to go for it :+1:

**Pull request merged**